### ISSUE TITLE
Refactor/fix boundary condition code

### ISSUE THREAD

**Issue by mirenradia** [GRTLCollaboration/GRTeclyn#62]
The boundary condition code is in a bit of a mess. I haven't tested anything beyond Sommerfeld and reflective conditions but I think these may be the only working non-periodic ones.

In the Sommerfeld case, we now apply the condition within the valid domain unlike for GRChombo (where we apply it in the ghost cells exterior to the valid domain). AMReX can handle extrapolating conditions but this is then applied in the exterior ghost cells (we actually do this currently in the Sommerfeld case to avoid NaNs not that it should affect anything). Is this OK or, for consistency, if using mixed BCs, do we need to apply the extrapolation in the same cells as for Sommerfeld? What do you think @KAClough?

The existing code should also be refactored/renamed to make it clear these are separate and in addition to AMReX's boundary condition code.

**Comment by KAClough** [comment]
Yes this is a good point. For the mixed BCs it would need to be consistent. I think it would be nice to have the option to use the AMReX BCs but also I do like being able to have the framework to create our own bespoke ones. So I would say for ours we should consistently apply them within the domain.

**Comment by mirenradia** [comment]
The problem is that AMReX BCs can't be applied to the RHS which we need to do for static and Sommerfeld BCs.

**Comment by KAClough** [comment]
We should also make sure we implement the fix in https://github.com/GRTLCollaboration/GRChombo/pull/271 

and also consider including the correct factors of the lapse etc in Sommerfeld BCs as @JCAurre says this makes a difference.

**Comment by KAClough** [comment]
This has now been somewhat addressed in #215 but still needs work.

**Comment by SeppeStaelens** [comment]
Dropping a comment here for future reference, relating to the Sommerfeld BCs.

While testing an implementation of the constraint violation norms in the TeukolskyWave example that I am adding ([PR #252](https://github.com/GRTLCollaboration/GRTeclyn/pull/252)), I obtained the following plot for the (volume-weighted L2 norm of the) Hamiltonian constraint over time:

<img width="320" height="300" alt="Image" src="https://github.com/user-attachments/assets/7f628b74-af67-431e-a9c4-0c3f61d11741" />

The largest one is the value obtained in GRTeclyn; the blue one at the bottom is GRChombo. The purple one is GRTeclyn, but ignoring the three outer layers of the unrefined 192^3 grid. All curves overlap until $t\approx 50$, when the ingoing wave reaches the origin and the outgoing wave starts reaching the outer boundary (although this may take another $\Delta t\approx20$ for the peak of the wave to get there).

The following plot shows the constraint violation on the edge surfaces of the grid, clearly being dominated by the outer layer or two in which the BC are applied (note that this is an octant, and the origin is in the left-most vertex as can be seen from the axes).

<img width="350" height="280" alt="Image" src="https://github.com/user-attachments/assets/87ad8669-4a92-4eaf-a00e-b8e2e03634de" />

As @mirenradia pointed out in the first comment, this is the result of applying the Sommerfeld BCs within the domain, rather than in ghost layers around it. These violations near the boundary then dominate the volume-weighted norm, which does not happen in GRChombo.

This is so noticeable here because the spacetime is a small vacuum perturbation, meaning that there is not e.g. a BH or boson star to dominate the violations in the interior of the grid. This hence may not be much of an issue for e.g. the BinaryBH example, but worth flagging I think. It is also hard to gauge whether this actually affects the evolution much or not in those cases.

**Comment by mirenradia** [comment]
That's interesting. Maybe we want to add the option to [optionally] exclude these cells when doing #73?
